src/lib/io/sol-uart-impl-linux.c

```diff
@@ -32,49 +33,175 @@
 #include "sol-uart.h"
 #include "sol-util-internal.h"
 #include "sol-util-file.h"
+#include "sol-vector.h"
 
 SOL_LOG_INTERNAL_DECLARE_STATIC(_log_domain, "uart");
 
 #define FD_ERROR_FLAGS (SOL_FD_FLAGS_ERR | SOL_FD_FLAGS_HUP | SOL_FD_FLAGS_NVAL)
 
+#define DEFAULT_BUFFER_SIZE (4096)
+
 struct sol_uart {
+    struct sol_fd *fd_handler;
+    const void *user_data;
+    ssize_t (*rx_cb)(void *data, struct sol_uart *uart, const struct sol_buffer *buf);
+    void (*tx_cb)(void *data, struct sol_uart *uart, struct sol_blob *blob, int status);
+    struct sol_timeout *read_timeout;
+    struct sol_ptr_vector pending_blobs;
+    struct sol_buffer rx;
+    size_t tx_size;
+    size_t pending_tx;
+    size_t written; //How many bytes We've written so far for the first element of pending_blobs
     int fd;
-    struct {
-        void *rx_fd_handler;
-        void (*rx_cb)(void *data, struct sol_uart *uart, uint8_t byte_read);
-        const void *rx_user_data;
-
-        void *tx_fd_handler;
-        void (*tx_cb)(void *data, struct sol_uart *uart, uint8_t *tx, int status);
-        const void *tx_user_data;
-        const uint8_t *tx_buffer;
-        unsigned int tx_length, tx_index;
-    } async;
+    bool delete_me;
+    bool in_use;
 };
 
+static void
+close_uart(struct sol_uart *uart)
+{
+    struct sol_blob *blob;
+    uint16_t i;
+
+    if (uart->fd_handler) {
+        sol_fd_del(uart->fd_handler);
+        uart->fd_handler = NULL;
+    }
+
+    if (uart->read_timeout) {
+        sol_timeout_del(uart->read_timeout);
+        uart->read_timeout = NULL;
+    }
+
+    uart->in_use = true;
+    SOL_PTR_VECTOR_FOREACH_IDX (&uart->pending_blobs, blob, i) {
+        if (uart->tx_cb)
+            uart->tx_cb((void *)uart->user_data, uart, blob, -ECANCELED);
```

Comment:
I can set `delete_me` to `true` here as well. By doing so, `sol_uart_write()` will fail